File: python/src/producer.py
#!/usr/bin/env python

from random import choice
from confluent_kafka import Producer
from dotenv import load_dotenv
import os

load_dotenv()

if __name__ == '__main__':
    config = {
        'bootstrap.servers': os.getenv('BOOTSTRAP_SERVER_URL'),
        'sasl.username': os.getenv('CONFLUENT_CLUSTER_API_KEY'),
        'sasl.password': os.getenv('CONFLUENT_CLUSTER_API_SECRET'),

        'security.protocol': 'SASL_SSL',
        'sasl.mechanisms': 'PLAIN',
        'acks': 'all'
    }

    producer = Producer(config)

    # Optional per-message delivery callback (triggered by poll() or flush())
    # when a message has been successfully delivered or permanently
    # failed delivery (after retries).
    def delivery_callback(err, msg):
        if err:
            print('ERROR: Message failed delivery: {}'.format(err))
        else:
            print("Produced event to topic {topic}: key = {key:12} value = {value:12}".format(
                topic=msg.topic(), key=msg.key().decode('utf-8'), value=msg.value().decode('utf-8')
            ))
    
    topic = 'purchases'
    users = ['eabara', 'jsmith', 'sgarcia', 'jbernard', 'htanaka', 'awalther']
    products = ['book', 'alarm clock', 't-shirts', 'gift card', 'batteries']

    count = 0
    for _ in range(10):
        user = choice(users)
        product = choice(products)
        producer.produce(topic, user, product, callback=delivery_callback)
        count += 1
    
    # Block until the messages are sent.
    producer.poll(10000)
    producer.flush()

File: golang/src/producer.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"log"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
    "github.com/joho/godotenv"
)

func main() {
	err := godotenv.Load()
	if err != nil {
		log.Fatal("Error loading .env file")
	}

	bootstrapServerUrl := os.Getenv("BOOTSTRAP_SERVER_URL")
	confluentClusterApiKey := os.Getenv("CONFLUENT_CLUSTER_API_KEY")
	confluentClusterApiSecret := os.Getenv("CONFLUENT_CLUSTER_API_SECRET")
	
	p, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": bootstrapServerUrl,
		"sasl.username": confluentClusterApiKey,
		"sasl.password": confluentClusterApiSecret,

		"security.protocol": "SASL_SSL",
		"sasl.mechanisms": "PLAIN",
		"acks": "all",
	})

	if err != nil {
		fmt.Printf("Failed to create producer: %s\n", err)
		os.Exit(1)
	}

	// Go-routine to handle message delivery reports and
    // possibly other event types (errors, stats, etc)
	go func() {
		for e := range p.Events() {
			switch ev := e.(type) {
			case *kafka.Message:
				if ev.TopicPartition.Error != nil {
					fmt.Printf("Failed to deliver message: %v\n", ev.TopicPartition)
				} else {
					fmt.Printf("Produced event to topic %s: key = %-10s value = %s\n",
						*ev.TopicPartition.Topic, string(ev.Key), string(ev.Value))
				}
			}
		}
	}()

	users := [...]string{"eabara", "jsmith", "sgarcia", "jbernard", "htanaka", "awalther"}
	items := [...]string{"book", "alarm clock", "t-shirts", "gift card", "batteries"}
	topic := "purchases"

	for n := 0; n < 10; n++ {
		key := users[rand.Intn(len(users))]
		data := items[rand.Intn(len(items))]
		
		p.Produce(&kafka.Message{
			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
			Key: []byte(key),
			Value: []byte(data),
		}, nil)
	}

	// Wait for all messages to be delivered
	p.Flush(15 * 1000)
	p.Close()
}